osbenchmark/synthetic_data_generator/mapping_synthetic_data_generator.py

```diff
@@ -0,0 +1,438 @@
+    generated_dataset_details = []
+
+    logger.info("Average document size: %s", avg_document_size)
+    logger.info("Chunk size: %s docs", docs_per_chunk)
+    logger.info("Total GB to generate: %s", sdg_config.total_size_gb)
+    logger.info("Max file size in GB: %s", generation_settings.get('max_file_size_gb'))
+
+    console.println(f"Total GB to generate: {sdg_config.total_size_gb}\n"
+                    f"Average document size: {avg_document_size}\n"
+                    f"Max file size in GB: {generation_settings.get('max_file_size_gb')}\n")
+
+    start_time = time.time()
+    with tqdm(total=total_size_bytes,
+                unit='B',
+                unit_scale=True,
+                bar_format="{l_bar}{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}, {rate_fmt}]") as progress_bar:
+
+        setup_custom_tqdm_formatting(progress_bar)
+        while current_size < total_size_bytes:
+            file_path = os.path.join(sdg_config.output_path, f"{sdg_config.index_name}_{file_counter}.json")
+            file_size = 0
+            docs_written = 0
+
+            while file_size < max_file_size_bytes:
+                generation_start_time = time.time()
+                seeds = generate_seeds_for_workers(regenerate=True)
+                logger.info("Using seeds: %s", seeds)
+
+                # TODO: Submit the seeds to the client so that each client is producing different variations of documents
+                futures = [client.submit(MappingSyntheticDataGeneratorWorker.generate_documents_from_worker, index_mappings, mapping_config, docs_per_chunk) for seed in seeds]
+                results = client.gather(futures) # if using AS_COMPLETED remove this line
+
+                writing_start_time = time.time()
+                for data in results:
+                    written = write_chunk(data, file_path)
+                    docs_written += written
+                    written_size = written * avg_document_size
+                    current_size += written_size
+                    progress_bar.update(written_size)
+
+                writing_end_time = time.time()
+
+                file_size = os.path.getsize(file_path)
+                # If it exceeds the max file size, then append this to keep track of record
+                if file_size >= max_file_size_bytes:
+                    file_name = file_path.split("/")[-1]
+                    generated_dataset_details.append({
+                        "file_name": file_name,
+                        "docs": docs_written,
+                        "file_size_bytes": file_size
+                    })
+
+                generating_took_time = writing_start_time - generation_start_time
+                writing_took_time = writing_end_time - writing_start_time
+                logger.info("Generating took [%s] seconds", generating_took_time)
+                logger.info("Writing took [%s] seconds", writing_took_time)
+
+                if current_size >= total_size_bytes:
```

Comment:
For the last file, the entry will be duplicated here, right?

Reply:
There's a guarantee that the last file will not exceed max file size along with max data size. Here's an example of when I generated 100GB of data with 40GB chunks. The last file is 20GB.
```
-rw-rw-r--. 1 ec2-user ec2-user   40G Mar 27 19:33 fidelity-test-100gb_0.json
-rw-rw-r--. 1 ec2-user ec2-user   40G Mar 27 20:50 fidelity-test-100gb_1.json
-rw-rw-r--. 1 ec2-user ec2-user   20G Mar 27 21:28 fidelity-test-100gb_2.json
```